### PR TITLE
[Enhancement] Consider ttl_number during optimization refresh

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 18 / 21 (85.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|5|7|71.43%|[377, 381]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|9|10|90.00%|[594]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 1]
```diff
@@ -1104,6 +1104,15 @@ public Set<String> getPartitionNames() {
         return Sets.newHashSet(nameToPartition.keySet());
     }
 
+    public Set<String> getValidPartitionNames(int partitionTTLNumber) {
+        // ess than 0 means not set
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:1108]:
```suggestion
        // less than 0 means not set
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 2]
```diff
@@ -1104,6 +1104,15 @@ public Set<String> getPartitionNames() {
...
+        if (partitionTTLNumber < 0) {
+            return getPartitionNames();
+        }
+        int startIndex = Math.max(0, nameToPartition.size() - partitionTTLNumber);
+        return nameToPartition.keySet().stream().skip(startIndex).collect(Collectors.toSet());
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:1113]: is keySet sorted?

File: fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java [hunk 3]
```diff
@@ -354,30 +352,35 @@ public static Map<String, Range<PartitionKey>> diffRange(Map<String, Range<Parti
...
 
         // There is no need to check autoRefreshPartitionsLimit if it is not auto refresh
         if (noPartitionRange && !isAutoRefresh) {
-            return materializedView.getPartitionNames();
+            return materializedView.getValidPartitionNames(partitionTTLNumber);
         }
-
-        if (noPartitionRange && isAutoRefresh) {
+        // Implies isAutoRefresh is true
+        if (noPartitionRange) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java:370]: It looks like not compute partitionTTLNumber when noPartitionRange is fasle

**Comment by Youngwb** [comment]
After this enhancement, mv refresh task will create all partitions but not load data into the partitions which are not in ttl partition numers?

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java (hunk 2 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java]: keySet is sorted but from logical is wrong.

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 4]
```diff
@@ -1104,6 +1104,15 @@ public Set<String> getPartitionNames() {
         return Sets.newHashSet(nameToPartition.keySet());
     }
 
+    public Set<String> getValidPartitionNames(int partitionTTLNumber) {
+        // ess than 0 means not set
+        if (partitionTTLNumber < 0) {
+            return getPartitionNames();
+        }
+        int startIndex = Math.max(0, nameToPartition.size() - partitionTTLNumber);
```
[the remark this answers is not in the document]
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java]: yes, we need.

**Comment by Astralidea** [comment]
> After this enhancement, mv refresh task will create all partitions but not load data into the partitions which are not in ttl partition numers?

This will in next PR.

File: fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java [hunk 5]
```diff
@@ -354,51 +351,41 @@ public static Map<String, Range<PartitionKey>> diffRange(Map<String, Range<Parti
...
+                if (!(lowerCmp >= 0 || upperCmp <= 0)) {
+                    result.add(entry.getKey());
+                }
             }
-            LiteralExpr startExpr = sortedRange.get(partitionNum - autoRefreshPartitionsLimit).lowerEndpoint().
-                    getKeys().get(0);
-            LiteralExpr endExpr = sortedRange.get(partitionNum - 1).upperEndpoint().getKeys().get(0);
-            start = AnalyzerUtils.parseLiteralExprToDateString(startExpr, 0);
-            end = AnalyzerUtils.parseLiteralExprToDateString(endExpr, 0);
+            return result;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java:374]: Not use ttl number to compute?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 53 / 56 (94.64%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|9|10|90.00%|[594]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|23|24|95.83%|[1116]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|21|22|95.45%|[379]|

**Review by Youngwb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 6]
```diff
@@ -1098,6 +1101,42 @@ public Set<String> getPartitionNames() {
...
+        LiteralExpr startExpr = sortedRange.get(partitionNum - lastPartitionNum).lowerEndpoint().
+                getKeys().get(0);
+        LiteralExpr endExpr = sortedRange.get(partitionNum - 1).upperEndpoint().getKeys().get(0);
+        String start = AnalyzerUtils.parseLiteralExprToDateString(startExpr, 0);
+        String end = AnalyzerUtils.parseLiteralExprToDateString(endExpr, 0);
+
+        Map<String, Range<PartitionKey>> result = Maps.newHashMap();
+        Column partitionColumn = ((RangePartitionInfo) partitionInfo).getPartitionColumns().get(0);
+        Range<PartitionKey> rangeToInclude = SyncPartitionUtils.createRange(start, end, partitionColumn);
+        for (Map.Entry<String, Range<PartitionKey>> entry : rangePartitionMap.entrySet()) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:1128]: use treeset to get top-k map entry from rangePartitionMap into result?

**Review by ABingHuang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**